x/evm/abci.go

```diff
@@ -542,13 +543,103 @@ func handleConfirmedEvents(ctx sdk.Context, bk types.BaseKeeper, n types.Nexus,
 	}
 }
 
+func validateGeneralMessage(ctx sdk.Context, ck types.ChainKeeper, n types.Nexus, m types.MultisigKeeper, chain nexus.Chain, msg nexus.GeneralMessage) (sdk.Int, multisig.KeyID, error) {
+	chainID := funcs.MustOk(ck.GetChainID(ctx))
+
+	keyID, ok := m.GetCurrentKeyID(ctx, chain.Name)
+	if !ok {
+		return chainID, keyID, fmt.Errorf("current key not set")
+	}
+
+	if !n.IsChainActivated(ctx, chain) {
+		return chainID, keyID, fmt.Errorf("destination chain de-activated")
+	}
+
+	if len(msg.Receiver) != 0 && !common.IsHexAddress(msg.Receiver) {
+		return chainID, keyID, fmt.Errorf("invalid contract address")
+	}
+
+	if _, ok := ck.GetGatewayAddress(ctx); !ok {
+		return chainID, keyID, fmt.Errorf("destination chain gateway not deployed yet")
+	}
+	return chainID, keyID, nil
+}
+
+func handleGeneralMessage(ctx sdk.Context, ck types.ChainKeeper, n types.Nexus, m types.MultisigKeeper, chain nexus.Chain, msg nexus.GeneralMessage) (types.CommandID, error) {
+	chainID, keyID, err := validateGeneralMessage(ctx, ck, n, m, chain, msg)
+	if err != nil {
+		return types.CommandID{}, err
+	}
+
+	dummyTxID := common.BytesToHash(make([]byte, common.HashLength))
+	cmd := types.NewApproveContractCallCommandGeneric(chainID, keyID, common.HexToAddress(msg.Receiver), common.BytesToHash(msg.PayloadHash), dummyTxID, msg.SourceChain, msg.Sender, msg.ID.ID)
+	funcs.MustNoErr(ck.EnqueueCommand(ctx, cmd))
+
+	return cmd.ID, nil
+}
+
+func handleGeneralMessages(ctx sdk.Context, bk types.BaseKeeper, n types.Nexus, m types.MultisigKeeper) {
+
+	for _, chain := range slices.Filter(n.GetChains(ctx), types.IsEVMChain) {
+		ck := funcs.Must(bk.ForChain(ctx, chain.Name))
+		endBlockerLimit := ck.GetParams(ctx).EndBlockerLimit
+		msgs := n.GetApprovedMessages(ctx, chain.Name, endBlockerLimit)
+		bk.Logger(ctx).Debug(fmt.Sprintf("handling %x general messages", len(msgs)), types.AttributeKeyChain, chain.Name)
+
+		for _, msg := range msgs {
+			success := utils.RunCached(ctx, bk, func(ctx sdk.Context) (bool, error) {
+				cmdID, err := handleGeneralMessage(ctx, ck, n, m, chain, msg)
+				if err != nil {
+					ck.Logger(ctx).Debug(fmt.Sprintf("failed handling general message: %s", err.Error()),
+						types.AttributeKeyChain, chain.Name.String(),
+						types.AttributeKeyMessageID, msg.ID.ID,
+					)
+					events.Emit(ctx, &types.ContractCallFailed{
+						Chain:     chain.Name,
+						MessageID: msg.ID.ID,
+						Reason:    err.Error(),
+					})
+
+					return false, err
```

Comment:
this should be done below, as the method could also panic causing a failure

Reply:
I put it here so we could access the error. Though I guess we could just capture an `err` variable and assign to it. Though I'm not sure what the default value should be, in the case that the method panics and there is no error.